File: code/sdk/android/cloudsheet-Android-sdk/src/main/kotlin/com/opencloudsheet/model/workbook/OneDriveWorkBook.kt
package com.opencloudsheet.model.workbook

import android.util.Log
import com.google.gson.Gson
import com.opencloudsheet.auth.IAuthenticator
import com.opencloudsheet.model.metadata.SheetMetadata
import com.opencloudsheet.model.worksheet.IWorksheetRow
import com.opencloudsheet.model.worksheet.IWorkSheet
import com.opencloudsheet.model.worksheet.OneDriveWorkSheet
import com.opencloudsheet.constants.OneDriveClient
import com.opencloudsheet.constants.OneDriveConstants
import com.opencloudsheet.metadata.MetadataManager
import com.opencloudsheet.metadata.OneDriveWorkBookMetadataInfo
import com.opencloudsheet.response.onedrive.WorkSheetListResponse
import com.opencloudsheet.response.onedrive.CreateWorkSheetRequest
import com.opencloudsheet.response.onedrive.WorkSheetData
import com.opencloudsheet.response.onedrive.WorkSheetResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.MediaType.Companion.toMediaType

/**
 * Microsoft OneDrive implementation of IWorkBook supporting multiple sheet types.
 *
 * Provides operations on an Excel workbook stored in OneDrive:
 * - Create typed sheets with automatic schema generation
 * - List all sheets via metadata
 * - Get specific sheet by name and type
 * - Delete sheets
 *
 * @see <a href="https://learn.microsoft.com/en-us/graph/api/resources/workbookworksheet?view=graph-rest-1.0">Microsoft Graph: Worksheet</a>
 */
class OneDriveWorkBook(
    private val metadataInfo: OneDriveWorkBookMetadataInfo,
    private val authenticator: IAuthenticator,
    private val workBookEntry: MetadataManager.WorkBookEntry,
    private val oneDriveClient: OneDriveClient
) : IWorkBook {

    companion object {
        private const val TAG = "OneDriveWorkBook"
        private const val METADATA_SHEET_NAME = "_Metadata"
        private val gson = Gson()
    }

    private var metadataSheet: OneDriveWorkSheet<SheetMetadata>? = null

    override fun getId(): String = metadataInfo.fileId
    override fun getName(): String = workBookEntry.name
    override fun getWorkBookEntry(): MetadataManager.WorkBookEntry = workBookEntry

    override suspend fun initialize() {
        val worksheets = getAllWorkSheetsFromExcel()

        val existingMetadataSheet = worksheets.find { it.name == METADATA_SHEET_NAME }
        if (existingMetadataSheet != null) {
            Log.i(TAG, "Found existing _Metadata sheet")
            metadataSheet = OneDriveWorkSheet(
                metadataInfo,
                existingMetadataSheet,
                SheetMetadata::class.java,
                authenticator,
                oneDriveClient
            )
        } else {
            Log.i(TAG, "Creating new _Metadata sheet")
            val createdWorksheet = createWorkSheetInExcel(METADATA_SHEET_NAME)
            metadataSheet = OneDriveWorkSheet(
                metadataInfo,
                createdWorksheet,
                SheetMetadata::class.java,
                authenticator,
                oneDriveClient
            )
            // Make the metadata sheet very hidden
            setWorkSheetVisibility(createdWorksheet.id, "VeryHidden")
        }
    }

    // MARK: - IWorkBook Protocol Methods

    override suspend fun <T : IWorksheetRow> createSheet(
        type: Class<T>,
        name: String,
        description: String
    ): IWorkSheet<T> {
        require(name.isNotBlank()) { "Sheet name cannot be blank" }

        val metadataSheet = this.metadataSheet
            ?: throw IllegalStateException("Workbook not initialized. Call initialize() first.")

        val existingSheets = metadataSheet.get()
        if (existingSheets.any { it.sheetName == name }) {
            throw IllegalArgumentException("Sheet with name '$name' already exists")
        }

        val workSheetData = createWorkSheetInExcel(name)

        val worksheet = OneDriveWorkSheet(
            metadataInfo,
            workSheetData,
            type,
            authenticator,
            oneDriveClient
        )

        val tempInstance = gson.fromJson("{}", type)
        val providerMetadataInfo = """{"iosClassName":"${tempInstance.getIosClassName()}","androidClassName":"${tempInstance.getAndroidClassName()}"}"""

        val sheetMetadata = SheetMetadata(
            sheetName = name,
            worksheetId = workSheetData.id,
            description = description,
            providerMetadataInfo = providerMetadataInfo
        )
        metadataSheet.create(sheetMetadata)

        Log.i(TAG, "Created sheet '$name' with type ${type.simpleName}")
        return worksheet
    }

    override suspend fun getSheets(): List<SheetMetadata> {
        val metadataSheet = this.metadataSheet
            ?: throw IllegalStateException("Workbook not initialized. Call initialize() first.")

        return metadataSheet.get()
    }

    override suspend fun <T : IWorksheetRow> getSheet(name: String, type: Class<T>): IWorkSheet<T>? {
        val metadataSheet = this.metadataSheet
            ?: throw IllegalStateException("Workbook not initialized. Call initialize() first.")

        val sheets = metadataSheet.get()
        val sheetMetadata = sheets.find { it.sheetName == name } ?: return null

        val workSheetData = WorkSheetData(
            id = sheetMetadata.worksheetId,
            name = sheetMetadata.sheetName,
            position = 0
        )

        return OneDriveWorkSheet(
            metadataInfo,
            workSheetData,
            type,
            authenticator,
            oneDriveClient
        )
    }

    override suspend fun deleteSheet(name: String) {
        val metadataSheet = this.metadataSheet
            ?: throw IllegalStateException("Workbook not initialized. Call initialize() first.")

        val sheets = metadataSheet.get()
        val sheetMetadata = sheets.find { it.sheetName == name }
            ?: throw IllegalArgumentException("Sheet with name '$name' not found")

        deleteWorkSheetInExcel(sheetMetadata.worksheetId)
        metadataSheet.delete(sheetMetadata)

        Log.i(TAG, "Deleted sheet '$name'")
    }

    // MARK: - Internal Excel Operations

    private suspend fun getAllWorkSheetsFromExcel(): List<WorkSheetData> {
        val token = authenticator.getAuthToken()
            ?: throw IllegalStateException("No authentication token available")

        val url = getWorksheetOperationUrl()

        val request = Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer $token")
            .get()
            .build()

        return withContext(Dispatchers.IO) {
            oneDriveClient.instance.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    Log.e(TAG, "Failed to get worksheets: ${response.code} - ${response.message}")
                    throw Exception("Failed to get worksheets: HTTP ${response.code}")
                }

                val responseBody = response.body.string()
                gson.fromJson(responseBody, WorkSheetListResponse::class.java).value
            }
        }
    }

    private suspend fun createWorkSheetInExcel(sheetName: String): WorkSheetData {
        val token = authenticator.getAuthToken()
            ?: throw IllegalStateException("No authentication token available")

        val createUrl = getWorksheetOperationUrl()

        val createRequestBody = CreateWorkSheetRequest(sheetName)
        val createRequestJson = gson.toJson(createRequestBody).toRequestBody("application/json".toMediaType())

        val createRequest = Request.Builder()
            .url(createUrl)
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
            .post(createRequestJson)
            .build()

        return withContext(Dispatchers.IO) {
            oneDriveClient.instance.newCall(createRequest).execute().use { response ->
                if (!response.isSuccessful) {
                    Log.e(TAG, "Failed to create worksheet: ${response.code} - ${response.message}")
                    throw Exception("Failed to create worksheet: HTTP ${response.code}")
                }

                val responseBody = response.body.string()
                val workSheetResponse = gson.fromJson(responseBody, WorkSheetResponse::class.java)

                WorkSheetData(
                    id = workSheetResponse.id,
                    name = workSheetResponse.name,
                    position = workSheetResponse.position
                )
            }
        }
    }

    private suspend fun deleteWorkSheetInExcel(worksheetId: String) {
        val token = authenticator.getAuthToken()
            ?: throw IllegalStateException("No authentication token available")

        val url = "${getWorksheetOperationUrl()}/$worksheetId"

        val request = Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer $token")
            .delete()
            .build()

        withContext(Dispatchers.IO) {
            oneDriveClient.instance.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    Log.e(TAG, "Failed to delete worksheet: ${response.code} - ${response.message}")
                    throw Exception("Failed to delete worksheet: HTTP ${response.code}")
                }
            }
        }
    }

    private fun getBaseWorkbookUrl(): String {
        return "${OneDriveConstants.BASE_MS_GRAPH_URL}/users('${metadataInfo.ownerId}')/drive/items('${metadataInfo.fileId}')/workbook"
    }

    private fun getWorksheetOperationUrl(): String {
        return "${getBaseWorkbookUrl()}/worksheets"
    }

    private suspend fun setWorkSheetVisibility(worksheetId: String, visibility: String) {
        val token = authenticator.getAuthToken()
            ?: throw IllegalStateException("No authentication token available")

        val url = "${getWorksheetOperationUrl()}/$worksheetId"
        val requestBody = """{"visibility":"$visibility"}""".toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
            .patch(requestBody)
            .build()

        withContext(Dispatchers.IO) {
            oneDriveClient.instance.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    Log.e(TAG, "Failed to set worksheet visibility: ${response.code} - ${response.message}")
                    throw Exception("Failed to set worksheet visibility: HTTP ${response.code}")
                }
                Log.i(TAG, "Set worksheet $worksheetId visibility to $visibility")
            }
        }
    }
}

File: code/sdk/ios/OpenCloudSheet/Sources/OpenCloudSheet/Model/Workbook/OneDriveWorkBook.swift
//
//  OneDriveWorkBook.swift
//  OpenCloudSheet
//
//  Microsoft OneDrive implementation of IWorkBook supporting multiple sheet types
//

import Foundation
import OSLog

class OneDriveWorkBook: IWorkBook {
    private let metadataInfo: OneDriveWorkBookMetadataInfo
    private let authenticator: IAuthenticator
    private let workBookEntry: WorkBookEntry
    private let oneDriveClient: OneDriveClient
    private var metadataSheet: OneDriveWorkSheet<SheetMetadata>?
    private let logger = Logger(subsystem: "com.opencloudsheet", category: "OneDriveWorkBook")

    private static let METADATA_SHEET_NAME = "_Metadata"

    init(
        metadataInfo: OneDriveWorkBookMetadataInfo,
        authenticator: IAuthenticator,
        workBookEntry: WorkBookEntry,
        oneDriveClient: OneDriveClient
    ) {
        self.metadataInfo = metadataInfo
        self.authenticator = authenticator
        self.workBookEntry = workBookEntry
        self.oneDriveClient = oneDriveClient
    }

    func initialize() async throws {
        let worksheets = try await getAllWorkSheetsFromExcel()

        if let metadataWorksheet = worksheets.first(where: { $0.name == OneDriveWorkBook.METADATA_SHEET_NAME }) {
            logger.info("Found existing _Metadata sheet")
            self.metadataSheet = OneDriveWorkSheet<SheetMetadata>(
                metadataInfo: metadataInfo,
                workSheetData: metadataWorksheet,
                clazz: SheetMetadata.self,
                authenticator: authenticator,
                oneDriveClient: oneDriveClient
            )
        } else {
            logger.info("Creating new _Metadata sheet")
            let createdWorksheet = try await createWorkSheetInExcel(sheetName: OneDriveWorkBook.METADATA_SHEET_NAME)
            self.metadataSheet = OneDriveWorkSheet<SheetMetadata>(
                metadataInfo: metadataInfo,
                workSheetData: createdWorksheet,
                clazz: SheetMetadata.self,
                authenticator: authenticator,
                oneDriveClient: oneDriveClient
            )
            try await setWorkSheetVisibility(worksheetId: createdWorksheet.id, visibility: "VeryHidden")
        }
    }

    func getId() -> String {
        return metadataInfo.fileId
    }

    func getName() -> String {
        return workBookEntry.name
    }

    func getWorkBookEntry() -> WorkBookEntry {
        return workBookEntry
    }

    // MARK: - IWorkBook Protocol Methods

    func createSheet<T: IWorksheetRow>(type: T.Type, name: String, description: String) async throws -> any IWorkSheet<T> {
        guard !name.isEmpty else {
            throw NSError(domain: "OneDriveWorkBook", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "Sheet name cannot be blank"])
        }

        guard let metadataSheet = metadataSheet else {
            throw NSError(domain: "OneDriveWorkBook", code: -2,
                         userInfo: [NSLocalizedDescriptionKey: "Workbook not initialized. Call initialize() first."])
        }

        let existingSheets = try await metadataSheet.get()
        if existingSheets.contains(where: { $0.sheetName == name }) {
            throw NSError(domain: "OneDriveWorkBook", code: -3,
                         userInfo: [NSLocalizedDescriptionKey: "Sheet with name '\(name)' already exists"])
        }

        let workSheetData = try await createWorkSheetInExcel(sheetName: name)

        let worksheet = OneDriveWorkSheet<T>(
            metadataInfo: metadataInfo,
            workSheetData: workSheetData,
            clazz: type,
            authenticator: authenticator,
            oneDriveClient: oneDriveClient
        )

        let providerMetadataInfo = """
        {"iosClassName":"\(type.getIosClassName())","androidClassName":"\(type.getAndroidClassName())"}
        """

        let sheetMetadata = SheetMetadata(
            sheetName: name,
            worksheetId: workSheetData.id,
            description: description,
            providerMetadataInfo: providerMetadataInfo
        )
        _ = try await metadataSheet.create(row: sheetMetadata)

        logger.info("Created sheet '\(name)' with type \(String(describing: type))")
        return worksheet
    }

    func getSheets() async throws -> [SheetMetadata] {
        guard let metadataSheet = metadataSheet else {
            throw NSError(domain: "OneDriveWorkBook", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "Workbook not initialized. Call initialize() first."])
        }

        return try await metadataSheet.get()
    }

    func getSheet<T: IWorksheetRow>(name: String, type: T.Type) async throws -> (any IWorkSheet<T>)? {
        guard let metadataSheet = metadataSheet else {
            throw NSError(domain: "OneDriveWorkBook", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "Workbook not initialized. Call initialize() first."])
        }

        let sheets = try await metadataSheet.get()
        guard let sheetMetadata = sheets.first(where: { $0.sheetName == name }) else {
            return nil
        }

        let workSheetData = WorkSheetData(
            id: sheetMetadata.worksheetId,
            name: sheetMetadata.sheetName,
            position: 0
        )

        return OneDriveWorkSheet<T>(
            metadataInfo: metadataInfo,
            workSheetData: workSheetData,
            clazz: type,
            authenticator: authenticator,
            oneDriveClient: oneDriveClient
        )
    }

    func deleteSheet(name: String) async throws {
        guard let metadataSheet = metadataSheet else {
            throw NSError(domain: "OneDriveWorkBook", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "Workbook not initialized. Call initialize() first."])
        }

        let sheets = try await metadataSheet.get()
        guard let sheetMetadata = sheets.first(where: { $0.sheetName == name }) else {
            throw NSError(domain: "OneDriveWorkBook", code: -2,
                         userInfo: [NSLocalizedDescriptionKey: "Sheet with name '\(name)' not found"])
        }

        try await deleteWorkSheetInExcel(worksheetId: sheetMetadata.worksheetId)
        _ = try await metadataSheet.delete(row: sheetMetadata)

        logger.info("Deleted sheet '\(name)'")
    }

    // MARK: - Internal Excel Operations

    private func getAllWorkSheetsFromExcel() async throws -> [WorkSheetData] {
        guard let token = try await authenticator.getAuthToken() else {
            throw NSError(domain: "OneDriveWorkBook", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "No authentication token available"])
        }

        let url = URL(string: getWorksheetOperationUrl())!
        var request = URLRequest(url: url)
        request.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await oneDriveClient.session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            let errorBody = String(data: data, encoding: .utf8) ?? ""
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            logger.error("Failed to get worksheets: HTTP \(statusCode) - \(errorBody)")
            throw NSError(domain: "OneDriveWorkBook", code: -2,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to get worksheets"])
        }

        let workSheetListResponse = try JSONDecoder().decode(WorkSheetListResponse.self, from: data)
        return workSheetListResponse.value
    }

    private func createWorkSheetInExcel(sheetName: String) async throws -> WorkSheetData {
        guard let token = try await authenticator.getAuthToken() else {
            throw NSError(domain: "OneDriveWorkBook", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "No authentication token available"])
        }

        let createUrl = URL(string: getWorksheetOperationUrl())!
        var createRequest = URLRequest(url: createUrl)
        createRequest.httpMethod = "POST"
        createRequest.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        createRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")

        let createRequestBody = CreateWorkSheetRequest(name: sheetName)
        createRequest.httpBody = try JSONEncoder().encode(createRequestBody)

        let (data, response) = try await oneDriveClient.session.data(for: createRequest)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 201 else {
            let errorBody = String(data: data, encoding: .utf8) ?? ""
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            logger.error("Failed to create worksheet: HTTP \(statusCode) - \(errorBody)")
            throw NSError(domain: "OneDriveWorkBook", code: -2,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to create worksheet"])
        }

        let workSheetResponse = try JSONDecoder().decode(WorkSheetResponse.self, from: data)

        return WorkSheetData(
            id: workSheetResponse.id,
            name: workSheetResponse.name,
            position: workSheetResponse.position
        )
    }

    private func deleteWorkSheetInExcel(worksheetId: String) async throws {
        guard let token = try await authenticator.getAuthToken() else {
            throw NSError(domain: "OneDriveWorkBook", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "No authentication token available"])
        }

        let url = URL(string: "\(getWorksheetOperationUrl())/\(worksheetId)")!
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        request.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await oneDriveClient.session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 204 else {
            let errorBody = String(data: data, encoding: .utf8) ?? ""
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            logger.error("Failed to delete worksheet: HTTP \(statusCode) - \(errorBody)")
            throw NSError(domain: "OneDriveWorkBook", code: -2,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to delete worksheet"])
        }
    }

    private func setWorkSheetVisibility(worksheetId: String, visibility: String) async throws {
        guard let token = try await authenticator.getAuthToken() else {
            throw NSError(domain: "OneDriveWorkBook", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "No authentication token available"])
        }

        let url = URL(string: "\(getWorksheetOperationUrl())/\(worksheetId)")!
        var request = URLRequest(url: url)
        request.httpMethod = "PATCH"
        request.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")

        let requestBody = ["visibility": visibility]
        request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)

        let (data, response) = try await oneDriveClient.session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            let errorBody = String(data: data, encoding: .utf8) ?? ""
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            logger.error("Failed to set worksheet visibility: HTTP \(statusCode) - \(errorBody)")
            throw NSError(domain: "OneDriveWorkBook", code: -2,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to set worksheet visibility"])
        }

        logger.info("Set worksheet '\(worksheetId)' visibility to '\(visibility)'")
    }

    private func getBaseWorkbookUrl() -> String {
        return "\(OneDriveConstants.baseMSGraphURL)/users('\(metadataInfo.ownerId)')/drive/items('\(metadataInfo.fileId)')/workbook"
    }

    private func getWorksheetOperationUrl() -> String {
        return "\(getBaseWorkbookUrl())/worksheets"
    }
}
